Block Entities - Containers

natri0 opened the pull request:
Added an article about implementing `Container` in block entities. Some stuff (e.g. `ImplementedContainer`) copied over from [wiki's "tutorial:inventory"](https://wiki.fabricmc.net/tutorial:inventory).

Tutorial covers using the `ImplementedContainer` & implementing `WorldlyContainer`, as well as some basic user interaction (inserting items on right-click).

natri0 committed:
Add 'Block Inventories' article

Basically the same content as the article on the fabric wiki: https://wiki.fabricmc.net/tutorial:inventory

Uses the same `ImplementedInventory` class as that one to implement `Inventory` on the custom block entity

Now updated for latest!

natri0 committed:
Spotless & Checkstyle

Maybe we should get a Checkstyle rule to allow comments after blocks? It's making me do stuff like this:

// :::1
if (condition) {
    // code...
}

// :::1

...which is IMO kinda ugly

its-miroma commented:
Thank you for your contribution!

We are temporarily not accepting PRs until we finish the update process. Follow #242 for more information.

You may keep this PR open.

IMB11 commented:
Hi! The documentation has been updated to version 1.21.4. Please rebase your PR and update your code to align with the latest version. 

Alternatively, you can create a new branch based on the latest commit from the [`FabricMC:main`](https://github.com/FabricMC/fabric-docs/tree/main) branch, copy over your changes from the old [`natri0:inventory`](https://github.com/natri0/fabric-docs/tree/inventory) branch, update it, and open a new PR - this approach is often simpler than dealing with complex git commands if you're not the best git wrangler.

Once you've completed the updates and are confident everything is ready, feel free to unmark this as a draft. Thanks again for your contribution!

natri0 commented:
updated for 1.21.4 :)

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+<VideoPlayer src="/assets/develop/blocks/inventory_2.mp4" />
+
+## Sided Inventories
```
```suggestion
## Sided Inventories {#sided-inventories}
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+Now, when you right-click the block with an item, you'll no longer have an item! If you run `/data get block` on the block, you'll see the item in the `Items` field in the NBT.
+
+![Duplicator block & /data get block output showing the item in the inventory](/assets/develop/blocks/inventory_1.png)
```
```suggestion
![Duplicator block and output of `/data get block` showing the item in the inventory](/assets/develop/blocks/inventory_1.png)
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+Here, if the player is holding an item and there is an empty slot, we move the item from the player's hand to the block's inventory and return `ItemActionResult.SUCCESS`.
+
+Now, when you right-click the block with an item, you'll no longer have an item! If you run `/data get block` on the block, you'll see the item in the `Items` field in the NBT.
```
```suggestion
Now, when you right-click the block with an item, you'll no longer have it! If you run `/data get block` on the block, you'll see the item in the `Items` field in the NBT.
```

JR1811 commented on develop/blocks/block-containers.md:
```diff
@@ -0,0 +1,80 @@
+
+@[code transcludeWith=:::3](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/DuplicatorBlockEntity.java)
+
+The `DuplicatorBlock` should now have a `getTicker` method that returns a reference to `DuplicatorBlockEntity::tick`.
```
Do we really need a ticking functionality here?
From what I can see, it's just copying the slot and is returning it asap. 

So this could be handled in the block interaction method in the custom block class, right?

Or do something like duplicating the slot every 10 ticks and then storing it in the inventory, until it's being retrieved or the stack has hit its max size.

natri0 replied:
@JR1811 changed it to only output one item every second.

JR1811 commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+# Block Inventories {#inventories}
+
+In Minecraft, all blocks that can store items have an `Inventory`. This includes blocks like chests, furnaces, and hoppers.
```
Might be just me, but this kind of makes it sound like the `Inventory` interface is always needed to store any kind of ItemStack. Maybe mention it as "best practice" or something?

JR1811 reviewed (commented):
This looks pretty good so far!

I think it would be interesting to also expand this page to explain how storing things like ItemStacks, Fluids, Energy, and so on, could be stored and retrieved. Also on that note, an example on how to use the Fabric Transfer API might be useful for that too. Or this could get its extra page, if necessary.

its-miroma commented:
Hello @natri0, are you still working on this?

natri0 commented:
@its-miroma sorry for the delay, was busy; i'll get back to it next week :)

natri0 committed:
Fix compilation error

natri0 committed:
Requested changes

some wording stuff and also redid the DuplicatorBlockEntity ticker so that it has a reason to exist.

natri0 commented:
rebased to latest main and applied changes requested by @its-miroma.

> Also on that note, an example on how to use the Fabric Transfer API might be useful for that too. Or this could get its extra page, if necessary.

@JR1811 i think a separate page (maybe "under" this one like Block Entities -> Block Entity Renderers) and a note about the existence of Fabric Transfer API would be better? don't want it to get too long/complex with marginally related parts all in the same article.

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,83 @@
+---
+title: Inventories
```
```suggestion
title: Block Inventories
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,83 @@
+  - natri0
+---
+
+# Block Inventories {#inventories}
```
```suggestion
---
```

its-miroma commented on develop/blocks/block-containers.md:
```diff
@@ -0,0 +1,83 @@
+
+<VideoPlayer src="/assets/develop/blocks/inventory_2.mp4" />
+
+
```
```suggestion
```

its-miroma commented on reference/latest/src/main/java/com/example/docs/block/entity/ModBlockEntities.java:
```diff
@@ -12,12 +12,17 @@
 import com.example.docs.block.entity.custom.CounterBlockEntity;
+import com.example.docs.block.entity.custom.DuplicatorBlockEntity;
 import com.example.docs.block.entity.custom.EngineBlockEntity;
+import com.example.docs.block.entity.custom.DuplicatorBlockEntity;
```
```suggestion
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,83 @@
+
+The `DuplicatorBlock` should now have a `getTicker` method that returns a reference to `DuplicatorBlockEntity::tick`.
+
+<VideoPlayer src="/assets/develop/blocks/inventory_2.mp4" />
```
```suggestion
<VideoPlayer src="/assets/develop/blocks/inventory_2.mp4">Duplicator block duplicating an oak log</VideoPlayer>
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,83 @@
+
+Now, if we try to insert items from the side instead of the top, it won't work!
+
+<VideoPlayer src="/assets/develop/blocks/inventory_3.webm" />
```
```suggestion
<VideoPlayer src="/assets/develop/blocks/inventory_3.webm">The Duplicator only activating when interacting with its top side</VideoPlayer>
```

its-miroma commented on public/assets/develop/blocks/inventory_2.mp4:
I don't know if it's on the guidelines, but we're only using `webm`

its-miroma reviewed (changes requested):
lint fixes

natri0 committed:
Apply (some) suggestions

only one missing is .mp4 -> .webm, need to reencode.

Co-authored-by: Miroma <136986257+its-miroma@users.noreply.github.com>

natri0 committed:
Reencode inventory_2.mp4 to webm

natri0 committed:
missed a suggestion

Co-authored-by: Miroma <136986257+its-miroma@users.noreply.github.com>

JR1811 commented on .vitepress/sidebars/develop.ts:
```diff
@@ -116,7 +116,11 @@ export default [
-    ],
+      {
+        text: "develop.blocks.inventory",
+        link: "/develop/blocks/inventory",
```
I would put Block Inventories under `Block Entities`.
Technically, a block(pos) inventory could be stored without a Block Entity using Persistent State.

So specifically ordering it to be categorized under `Block Entity` instead of `Block` might be clearing up that issue.

JR1811 commented on sidebar_translations.json:
```diff
@@ -35,6 +35,7 @@
   "develop.blocks.blockstates": "Block States",
   "develop.blocks.block-entities": "Block Entities",
   "develop.blocks.block-entity-renderer": "Block Entity Renderers",
+  "develop.blocks.inventory": "Block Inventories",
```
Change the address to reflect the entry being listed in `Blocks > Block Entity` instead of `Blocks`

(For more information, see requested changes in develop.ts file).

JR1811 commented on reference/latest/src/main/java/com/example/docs/block/entity/custom/DuplicatorBlockEntity.java:
```diff
@@ -0,0 +1,102 @@
+
+	private final DefaultedList<ItemStack> items = DefaultedList.ofSize(1, ItemStack.EMPTY);
+
+	// :::3
```
Reference comments are displayed in the code view?

![grafik](https://github.com/user-attachments/assets/02e716b1-4408-4fc3-bd34-c2f6f21d8c7c)

JR1811 reviewed (changes requested):
Small changes, nothing major 👍🏻

its-miroma commented:
Hi @natri0, sorry for the long wait, do you think you could address the last few reviews, and check if anything changed in 1.21.10, so we can get this merged? Thank you!

cassiancc committed:
Merge branch 'main' into pr/246

cassiancc committed:
Port to 1.21.10 and Mojang Mappings

Tenneb22 commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+@[code transcludeWith=:::accept](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/DuplicatorBlockEntity.java)
+
+The `getInvAvailableSlots` returns an array of the slot _indices_ that can be interacted with from the given side. In this case, we only have a single slot (`0`), so we return an array with just that index.
```
```suggestion
The `getSlotsForFace` returns an array of the slot _indices_ that can be interacted with from the given side. In this case, we only have a single slot (`0`), so we return an array with just that index.
```

Tenneb22 commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+The `getInvAvailableSlots` returns an array of the slot _indices_ that can be interacted with from the given side. In this case, we only have a single slot (`0`), so we return an array with just that index.
+
+Also, we should modify the `onUseWithItem` method of the `DuplicatorBlock` to actually respect the new behavior:
```
```suggestion
Also, we should modify the `useItemOn` method of the `DuplicatorBlock` to actually respect the new behavior:
```

Tenneb22 commented on reference/latest/src/main/java/com/example/docs/block/entity/custom/DuplicatorBlockEntity.java:
```diff
@@ -0,0 +1,102 @@
+
+	// :::tick
+	public static void tick(Level world, BlockPos blockPos, BlockState blockState, DuplicatorBlockEntity duplicatorBlockEntity) {
+		if (!duplicatorBlockEntity.isEmpty()) {
+			duplicatorBlockEntity.timeSinceDropped++;
+
+			if (duplicatorBlockEntity.timeSinceDropped < 10) return;
+			duplicatorBlockEntity.timeSinceDropped = 0;
+
+			ItemStack duplicate = duplicatorBlockEntity.getItem(0).split(1);
+
+			Block.popResourceFromFace(world, blockPos, Direction.UP, duplicate);
+			Block.popResourceFromFace(world, blockPos, Direction.UP, duplicate);
+		}
+	}
```
```suggestion
	public static void tick(Level world, BlockPos blockPos, BlockState blockState, DuplicatorBlockEntity duplicatorBlockEntity) {
		if (duplicatorBlockEntity.isEmpty()) return;
		duplicatorBlockEntity.timeSinceDropped++;

		if (duplicatorBlockEntity.timeSinceDropped < 10) return;
		duplicatorBlockEntity.timeSinceDropped = 0;

		ItemStack duplicate = duplicatorBlockEntity.getItem(0).split(1);

		Block.popResourceFromFace(world, blockPos, Direction.UP, duplicate);
		Block.popResourceFromFace(world, blockPos, Direction.UP, duplicate);
	}
```
Maybe use an early return for a cleaner code style?

cassiancc committed:
Apply suggestions from code review

Co-authored-by: Tenneb22 <126985099+Tenneb22@users.noreply.github.com>

cassiancc committed:
Fix data generation

cassiancc committed:
Merge branch 'main' into pr/246

Tenneb22 commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,80 @@
+
+@[code transcludeWith=:::block](@/reference/latest/src/main/java/com/example/docs/block/custom/DuplicatorBlock.java)
+
+Then, we need to create a `DuplicatorBlockEntity`, which needs to implement the `Container` interface. Thankfully, there's a helper called `ImplementedInventory` that does most of the work, leaving us with just a few methods to implement.
```
>  Thankfully, there's a helper called `ImplementedInventory` that does most of the work, leaving us with just a few methods to implement.

The tutorial mentions `ImplementedInventory`, but the code is never shown. I'd say that this could mislead someone into thinking it is a helper from Minecraft itself or fapi, which is not the case.

Is it feasible to embed the code for ImplementInventory, but make it collapsible, so it doesn't take up too much space on the page? ( something like `<details>`, but that doesn't work directly with a code reference )

its-miroma replied:
Yes, one can do:

```md
::: details Show `ImplementedInventory`

@[code transcludeWith=:::whatever](@/reference/.../ImplementedInventory.java)

:::
```

though as you said the sentence implies it comes from Minecraft or API, which must be clarified

cassiancc committed:
Add `ImplementedContainer` example

its-miroma commented on .vitepress/sidebars/develop.ts:
```diff
@@ -116,6 +116,10 @@ export default [
             link: "/develop/blocks/block-entity-renderer",
           },
+          {
+            text: "develop.blocks.inventory",
```
As was pointed out by JR1811, this should be under Block Entities?
Also, the file name can be `containers`.

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,86 @@
+---
+title: Block Containers
+description: Learn how to add containers to your blocks.
```
```suggestion
description: Learn how to add containers to your block entities.
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,86 @@
+  - natri0
+---
+
+In Minecraft, a best practice when creating blocks that can store items is to implement `Container`. It is implemented by blocks like chests and furnaces. This makes it possible to, for example, interact with the block using hoppers.
```
```suggestion
It's good practice when creating blocks that can store items, like chests and furnaces, is to implement `Container`. This makes it possible to, for example, interact with the block using hoppers.
```

its-miroma commented on develop/blocks/inventory.md:
```diff
@@ -0,0 +1,86 @@
+
+### Duplicating Items {#duplicating-items}
+
+Actually, on second thought, shouldn't a _duplicator_ block duplicate items? Let's make it so that the block duplicates the stack you threw in it, but only two items at a time. And let's make it wait a second every time to not spam the player with items!
```
```suggestion
Let's now make it so that the block duplicates the stack you threw in it, but only two items at a time. And let's make it wait a second every time to not spam the player with items!
```

its-miroma commented on reference/latest/src/main/java/com/example/docs/container/ImplementedContainer.java:
```diff
@@ -0,0 +1,135 @@
+ */
+public interface ImplementedContainer extends Container {
+	/**
+	 * Retrieves the item list of this inventory.
```
All comments in this file need to be updated to mojmap.

its-miroma reviewed (changes requested):
I think this article is very close to being merged. why was it still a draft?

cassiancc committed:
Apply suggestions from code review

Co-authored-by: Miroma <its.miroma@proton.me>

cassiancc committed:
Further containerification

cassiancc committed:
Update block-containers.md

its-miroma commented on develop/blocks/block-containers.md:
```diff
@@ -0,0 +1,86 @@
+  - natri0
+---
+
+It's good practice when creating blocks that can store items, like chests and furnaces, is to implement `Container`. This makes it possible to, for example, interact with the block using hoppers.
```
```suggestion
It's good practice when creating blocks that can store items, like chests and furnaces, to implement `Container`. This makes it possible to, for example, interact with the block using hoppers.
```

cassiancc committed:
Update develop/blocks/block-containers.md

Co-authored-by: Miroma <its.miroma@proton.me>
